Core: Revert Emotion `:first-child` (etc) workarounds

tmeasday opened the pull request:
Closes #20730

## What I did

As discussed [here](https://github.com/storybookjs/storybook/issues/20730#issuecomment-1439747642), I reverted https://github.com/storybookjs/storybook/pull/18361.

In the end there were no other styles that needed to be escaped.

## How to test

Run a sandbox, check both story + docs rendering.

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

tmeasday committed:
Revert "Drop `ignoreSsrWarning` workaround"

This reverts commit ca86e9d3900a342d48a624d8584bd57f544ad0d5.

tmeasday committed:
Revert "Turn off Emotion's warnings about potentially unsafe pseudo-selectors in SSR"

This reverts commit 03c5026744a4988fdc3623c1494d6efe58ef786d.

tmeasday commented:
@Andarist I am still seeing a problem on this branch. It seems like nested usage of the comment is not working (I'll call out the line of code that's causing the issue):

<img width="994" alt="image" src="https://user-images.githubusercontent.com/132554/220828865-5bb3bb77-ed4b-42f0-aa70-b1345e21780f.png">

tmeasday commented on code/ui/blocks/src/components/ArgsTable/ArgsTable.tsx:
```diff
@@ -112,20 +112,20 @@ export const TableWrapper = styled.table<{
-      [`tr:first-child`]: {
-        [`td:first-child, th:first-child`]: {
+      [`tr:first-child${ignoreSsrWarning}`]: {
+        [`td:first-child${ignoreSsrWarning}, th:first-child${ignoreSsrWarning}`]: {
```
These inner usages of the selector trigger the warning.

ndelangen commented:
@tmeasday let's discuss, when you're available?

Andarist commented:
Oh, I missed this notification last week - sorry!. Could you give me repro steps so I could jump directly into investigating this?

tmeasday commented:
Hey @Andarist sure thing, here are the instructions:

1. Check out this branch: `fix-emotion-warnings-for-good`
2. Run `yarn start` from the root
3. Browse to any story file, e.g. http://localhost:6006/?path=/story/example-button--primary
4. Also browse to any attached docs file, e.g. http://localhost:6006/?path=/docs/example-button--docs

I'm not sure if 3&4 are the same issue or not.

tmeasday commented:
@Andarist let me know if you need further help reproducing

ndelangen committed:
Merge branch 'next' into fix-emotion-warnings-for-good

ndelangen committed:
Merge branch 'next' into fix-emotion-warnings-for-good

Andarist commented:
Thanks for this I extracted a repro case for myself, it happens with those styles:
```css
&&{border-spacing:0;color:#C9CDCF;td, th{padding:0;border:none;vertical-align:top;text-overflow:ellipsis;}font-size:13px;line-height:20px;text-align:left;width:100%;margin-top:0;margin-bottom:0;thead th:first-of-type, td:first-of-type{width:25%;}th:first-of-type, td:first-of-type{padding-left:20px;}th:nth-of-type(2), td:nth-of-type(2){}td:nth-of-type(3){}th:last-of-type, td:last-of-type{padding-right:20px;}th{color:rgba(201,205,207,0.55);padding-top:10px;padding-bottom:10px;padding-left:15px;padding-right:15px;}td{padding-top:10px;padding-bottom:10px;&:not(:first-of-type){padding-left:15px;padding-right:15px;}&:last-of-type{padding-right:20px;}}margin-left:0;margin-right:0;tr:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{td:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-top-left-radius:0;}td:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-top-right-radius:0;}}tr:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{td:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-bottom-left-radius:0;}td:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-bottom-right-radius:0;}}tbody{tr{background:transparent;overflow:hidden;border-top-width:1px;border-top-style:solid;border-top-color:#27292a;}td{background:#1B1C1D;border-top:1px solid rgba(255,255,255,.1);}}}
```

Gonna investigate this - thanks for the ping.

tmeasday commented:
@Andarist any luck? We are hoping to release soon, I'm wondering if we should just fallback to the Cache Provider solution in docs.

Andarist commented:
I’ll get to it over the weekend

Andarist commented:
I've diagnosed the problem - now I "only" need to figure out the fix. I hate the code responsible for those warnings 🤣

Andarist committed:
Merge remote-tracking branch 'origin/next' into fix-emotion-warnings-for-good

Andarist committed:
Fixed comment placement

Andarist commented:
I fixed the issue in Emotion ([here](https://github.com/emotion-js/emotion/pull/3019)), bumped the `@emotion/cache` in this repo [here](https://github.com/storybookjs/storybook/pull/21941) and merged `next` into this branch. This fixed almost all of the issues.

One issue persisted though and I diagnosed it as a bug in `stylis` and I reported it [here](https://github.com/thysultan/stylis/issues/310). Luckily, it was easy to workaround it by just moving the comment outside of the parenthesis so I've done that and pushed it out.

Let me know if you find this warning being printed in any other case.

tmeasday commented:
Amazing @Andarist!

tmeasday commented:
@Andarist (or @ndelangen) can you approve (because I created the PR)?

tmeasday commented on code/ui/blocks/src/components/ArgsTable/ArgsTable.tsx:
```diff
@@ -107,6 +107,24 @@ export const TableWrapper = styled.table<{
       marginLeft: inAddonPanel ? 0 : 1,
       marginRight: inAddonPanel ? 0 : 1,
 
+      [`tr:first-child${ignoreSsrWarning}`]: {
+        [`td:first-child${ignoreSsrWarning}, th:first-child${ignoreSsrWarning}`]: {
+          borderTopLeftRadius: inAddonPanel ? 0 : theme.appBorderRadius,
+        },
+        [`td:last-child${ignoreSsrWarning}, th:last-child${ignoreSsrWarning}`]: {
+          borderTopRightRadius: inAddonPanel ? 0 : theme.appBorderRadius,
+        },
+      },
+
+      [`tr:last-child${ignoreSsrWarning}`]: {
+        [`td:first-child${ignoreSsrWarning}, th:first-child${ignoreSsrWarning}`]: {
+          borderBottomLeftRadius: inAddonPanel ? 0 : theme.appBorderRadius,
+        },
+        [`td:last-child${ignoreSsrWarning}, th:last-child${ignoreSsrWarning}`]: {
+          borderBottomRightRadius: inAddonPanel ? 0 : theme.appBorderRadius,
+        },
+      },
+
```
What's happened here? Where did all this code go/come from?

tmeasday replied:
Investigating...

Andarist replied:
huh, that's weird :D I'm pretty sure that I didn't do it... perhaps a wrong merge?

tmeasday replied:
Seems like some kind of merge issue, will fix.

tmeasday replied:
Fixed!

Andarist commented:
<img alt="a girl self-patting herself and saying 'you are welcome everyone'" src="https://media.giphy.com/media/gH3HHXL67YvoiewbND/giphy.gif"/>

tmeasday committed:
Fix bad merge in `ArgsTable`.

The relevant code was already removed on `next` in https://github.com/storybookjs/storybook/commit/5892ea4b15c448f5fe180e4e293eaea55196a257
